Mobile Keyboard Input

djclueless opened the pull request:
Here is a start for the mobile input portion, we need to work on key mapping though. Probably related but text is actually being typed into the input on mobile, that would need to be prevented.

Keyboard opens on default and full screen views. On android you can hide the keyboard with the back arrow, I have not tested on iOS.

I Dynamically created the input, but it can easily be included in the html.

---

**Edit by mtlynch@**

Fixes #46

^^^ Putting that line in the PR description auto-closes the linked issue when we merge it.

djclueless commented:
This if for issue #46

mtlynch commented on app/templates/custom-elements/remote-screen.html:
```diff
@@ -79,6 +79,20 @@
+              .appendChild(element);
+          }
+          element.focus();
+        }
```
In general, I think it's clearer to define HTML elements in HTML and CSS unless there's a good reason to avoid it.

djclueless replied:
Done

mtlynch commented on app/static/js/app.js:
```diff
@@ -325,6 +325,11 @@ document.getElementById("fullscreen-btn").addEventListener("click", (evt) => {
   document.getElementById("paste-overlay").show = true;
 });
+
+document.getElementById("remote-screen").addEventListener("click", () => {
```
Could the event registration live within `remote-screen.html`?

Also, would it make more sense to react to [touch events](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events/Using_Touch_Events) instead of mouse events?

djclueless replied:
Moved event to remote-screen.html but I couldn't get the touch events to work, the input gets focus but the keyboard wouldn't show for me.

mtlynch reviewed (commented):
Good start! Left a few early notes.

mtlynch commented on app/templates/custom-elements/remote-screen.html:
```diff
@@ -77,6 +83,24 @@
           });
+          let isTouchScreen = false;
+          this.shadowRoot.addEventListener("touchend", () => {
+            isTouchScreen = true;
```
Oh, this is clever! I wouldn't have thought of this. It might have some false positives at some point, but we can deal with it then.

mtlynch commented on app/templates/custom-elements/remote-screen.html:
```diff
@@ -77,6 +83,24 @@
+          this.shadowRoot
+            .getElementById("mobile-keyboard-input")
+            .addEventListener("input", (evt) => {
+              if (evt.target.style.display != "none") {
```
It doesn't look like the if clause is necessary since we never set display to `none`.

mtlynch commented on app/templates/custom-elements/remote-screen.html:
```diff
@@ -30,9 +30,15 @@
   </style>
   <div class="screen-wrapper">
     <img id="remote-screen-img" src="/stream?advance_headers=1" />
+    <input type="text" id="mobile-keyboard-input" />
```
On Android, it tries to autocapitalize on every keystroke because it's at the beginning of the text field. We can fix this by adding `autocapitalize="off"` to the `<input>` element.

mtlynch reviewed (changes requested):
Nice addition! I'm running into two bugs.

It's impossible to type backspaces. Could the input start with a space character, and then we detect "backspace" by checking to see if the text is `""`?

Then there's strange behavior with dashes. If I type dashes (`-`), they send normally, but for some reason, the next non-dash character I type includes all the previous dashes.

* `-l` sends `-` and then `-l`
* `--l` sends `-`, `-`, `--l`

Can you reproduce that behavior on your end? I'm on a Pixel2 + Chrome 85.

mtlynch commented on app/templates/custom-elements/remote-screen.html:
```diff
@@ -77,6 +83,24 @@
+            .getElementById("mobile-keyboard-input")
+            .addEventListener("input", (evt) => {
+              if (evt.target.style.display != "none") {
+                sendPastedText(evt.target.value);
```
It felt strange calling `sendPastedText` when we're not actually pasting text. I refactored the code (#288), so now it will work if you call `sendTextInput`.

djclueless commented:
you've been busy today, I just got home looking over the changes

djclueless commented:
> Nice addition! I'm running into two bugs.
> 
> It's impossible to type backspaces. Could the input start with a space character, and then we detect "backspace" by checking to see if the text is `""`?
> 
> Then there's strange behavior with dashes. If I type dashes (`-`), they send normally, but for some reason, the next non-dash character I type includes all the previous dashes.
> 
> * `-l` sends `-` and then `-l`
> * `--l` sends `-`, `-`, `--l`
> 
> Can you reproduce that behavior on your end? I'm on a Pixel2 + Chrome 85.

I can do both back space and dashes no issues

Moto G7 Android 10 
Chrome v85.0.4183.101

mtlynch commented:
>I can do both back space and dashes no issues

Hmm, that's puzzling. The current implementation reads the value of the `mobile-keyboard-input` and forwards it to the backend. An `<input>` element can't contain a backspace, so how does it detect the backspace?

Have you tried dash + another letter? Dashes work fine, but when you type a dash and then type a letter like `p` after the dash, it doubles the dashes.

djclueless commented:
I do see an issue with spaces and backspace. If I do 3 spaces I need to do 4 Backspaces for it to take. But if set an alert to notify when a key is pressed it works fine with the interruption.

I changed addEventListener("input" to "keyup" but still the same issue.

I didn't notice anything strange with the dash (-) I will mess with it some more tomorrow.

I converted it back to draft until we figure out what is happening.

mtlynch commented on app/static/js/app.js:
```diff
@@ -131,6 +131,10 @@ function isIgnoredKeystroke(keyCode) {
 // Send a keystroke message to the backend, and add a key card to the web UI.
 function sendKeystroke(keystroke) {
+  //Ignore if keyCode is 229 (mobile)
+  if (keystroke.keyCode === 229) {
```
Can we add more detail here about what keycode 229 is? Also, since we reference it in multiple places, can we define it as a named constant instead?

djclueless replied:
keyCode 229 is some kind of placeholder for a combined character, I think it has to do with the autocomplete in the mobile OS. It indicates something is sent but it's not quite done. I got around this by blurring the input essentially clearing the autocomplete and then re-focusing so we can accept the next character.

mtlynch commented on app/templates/custom-elements/remote-screen.html:
```diff
@@ -108,8 +118,28 @@
+          });
+          this.shadowRoot
+            .getElementById("mobile-keyboard-input")
+            .addEventListener("keyup", (evt) => {
```
Can we add a comment explaining this? I'm having trouble understanding the logic here. Also, how do we detect backspace? `innerHTML` can't be a backspace, can it?

djclueless replied:
Anything not 229 like backspace gets sent the original way.

mtlynch commented on app/templates/custom-elements/remote-screen.html:
```diff
@@ -108,8 +118,28 @@
+                if (evt.target.innerHTML == "&nbsp;") {
+                  evt.target.innerHTML = " ";
+                }
+                sendTextInput(evt.target.innerHTML);
```
Why `innerHTML` instead of `innerText`? 

The current implementation doesn't properly forward `<` or `>`. I think what we want is:

```javascript
if (evt.target.innerHTML === "&nbsp;") {
  sendTextInput(" ");
} else {
  sendTextInput(evt.target.innerText);
}
```

djclueless replied:
That doesn't fix anything its essentially doing the same thing. I can change it just so its clearer. It would be the same thing as doing the same thing below.

```
sendTextInput(evt.target.innerHTML === "&nbsp? " ": evt.target.innerText);
```

djclueless commented on app/static/css/style.css:
```diff
@@ -8,6 +8,10 @@ body {
 }
 
+#app {
+  display: grid;
```
on my phone the nav bar was not going all the way across. This had fixed it, but now I cant duplicate the problem lol.

We can take it out for now, if I run into the issue again I will do a screen capture and share with you.

djclueless replied:
Here you go, with and without.


![image](https://user-images.githubusercontent.com/22175951/93928954-1c43b400-fce9-11ea-864e-8947851b014d.png)

![image](https://user-images.githubusercontent.com/22175951/93929006-2d8cc080-fce9-11ea-8b17-2598971598e1.png)

mtlynch replied:
Ah, that's an unrelated issue then. I've filed #263 to track it. At first glance, it seems like a problem with the keyboard history panel rather than the `#app` CSS rules. I want to be conservative about setting CSS rules in the root element for the page.

mtlynch reviewed (approved):
LGTM, thanks!

djclueless commented:
hmm with these changes only alpha numeric characters work for me. Space, comma, period, and all the other characters don't. They show up as undefined. Also full screen on a desktop pc stopped working now, not sure what caused it to break.

mtlynch commented:
Whoops, reverting in #264.
